File: front/go/deepx/linear.go
package deepx

type Linear struct {
	ModuleBase
	W *Tensor
	b *Tensor
}

func NewLinear(name string, in_features, out_features int, dtype Dtype, g *Graph) (m *Linear) {
	if g == nil {
		g = NewGraph()
	}
	if name == "" {
		name = "linear"
	}
	m = &Linear{
		ModuleBase: ModuleBase{
			g:    g,
			name: name,
		},
	}

	in_features_node := g.AddConstArg(name + ".in_features")
	in_features_node.SetInt(in_features)
	out_features_node := g.AddConstArg(name + ".out_features")
	out_features_node.SetInt(out_features)

	//如果利用矩阵grad时的取巧运算，则需要将W的shape设置为[out_features,in_features]来实现提前转置
	m.W = g.AddTensor(name+".W", dtype, []int{in_features, out_features}, true, in_features_node, out_features_node).Tensor()
	m.b = g.AddTensor(name+".bias", dtype, []int{out_features}, true, out_features_node).Tensor()
	return m
}
func (m *Linear) Linear(input *Tensor) *Tensor {
	y := input.Matmul(m.W)
	z := y.Add(m.b)
	return z
}

File: front/py/deepx/nn/modules/linear.py
from .module import Module
from deepx  import Tensor
from deepx.nn.functional import uniform_,kaiming_uniform_,calculate_fan_in_and_fan_out
import math

class Linear(Module):
    r'''
    copy from torch.nn.Linear

    math:`y = xA^T + b` 
    not y=Ax+b
    '''
    def __init__(
        self,
        in_features: int,
        out_features: int,
        bias: bool = True,
        dtype:str="float32",
    ) -> None:
        super().__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.weight = Tensor(shape=(out_features, in_features),dtype=dtype)
        self.register_parameter("weight",self.weight)
        if bias:
            self.bias = Tensor(shape=(out_features,),dtype=dtype)
            self.register_parameter("bias",self.bias)
        else:
            self.register_parameter("bias", None)
        self.reset_parameters()

    def reset_parameters(self) -> None:
        # Setting a=sqrt(5) in kaiming_uniform is the same as initializing with
        # uniform(-1/sqrt(in_features), 1/sqrt(in_features)). For details, see
        # https://github.com/pytorch/pytorch/issues/57109
        kaiming_uniform_(self.weight, a=math.sqrt(5))
        if self.bias is not None:
            fan_in, _ = calculate_fan_in_and_fan_out(self.weight)
            bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
            uniform_(self.bias, -bound, bound)

    def forward(self, input: Tensor) -> Tensor:
        #`y = xA^T + b`
        y=input @ self.weight.T
        oldshape=y.shape
        if self.bias is not None:
            y.reshape_(tuple(y.shape[1:]))
            y=y+self.bias
        y.reshape_(oldshape)
        return y

    def extra_repr(self) -> str:
        return f"in_features={self.in_features}, out_features={self.out_features}, bias={self.bias is not None}"

